File: LeetCode/141-LinkedListCycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */

const hasCycle = (head) => {
    const values = {};

    let node = head;

    if (!node) return false;
    if (!node.next) return false;
    if (node.next.next === node) return true;

    while (node.next.next) {
        if (values[node.next.val] === node.next.next.val) {
            console.log(node.next.val)
            return true;
        }

        values[node.val] = node.next.val;

        node = node.next;
    }

    return false
};

File: leetcode_python/linked_list/141_linked_list_cycle.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """

        list_set = set()

        while head:
            if head in list_set:
                return True
            else:
                list_set.add(head)
                head = head.next

        return False
